fleetconfig-controller/internal/controller/addon.go

```diff
@@ -0,0 +1,218 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"slices"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	addonv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	addonapi "open-cluster-management.io/api/client/addon/clientset/versioned"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/open-cluster-management-io/lab/fleetconfig-controller/api/v1alpha1"
+	exec_utils "github.com/open-cluster-management-io/lab/fleetconfig-controller/internal/exec"
+	"github.com/open-cluster-management-io/lab/fleetconfig-controller/internal/file"
+)
+
+const (
+	addonConfigMapName = "fleet-add-ons"
+	addon              = "addon"
+	create             = "create"
+	enable             = "enable"
+	disable            = "disable"
+)
+
+type manifestType string
+
+const (
+	manifestTypeRaw  manifestType = "raw"
+	manifestTypeHTTP manifestType = "http"
+)
+
+func handleAddonConfig(ctx context.Context, kClient client.Client, addonC *addonapi.Clientset, fc *v1alpha1.FleetConfig) error {
+	logger := log.FromContext(ctx)
+	logger.V(0).Info("handleAddOnConfig", "fleetconfig", fc.Name)
+
+	// get existing addons
+	createdAddOns, err := addonC.AddonV1alpha1().AddOnTemplates().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	requestedAddOns := fc.Spec.AddOnConfigs
+
+	// nothing to do
+	if len(requestedAddOns) == 0 && len(createdAddOns.Items) == 0 {
+		logger.V(5).Info("no addons to reconcile")
+		return nil
+	}
+
+	// compare existing to requested
+	createdVersionedNames := make([]string, len(createdAddOns.Items))
+	for i, ca := range createdAddOns.Items {
+		createdVersionedNames[i] = ca.Name
+	}
+
+	requestedVersionedNames := make([]string, len(requestedAddOns))
+	for i, ra := range requestedAddOns {
+		requestedVersionedNames[i] = fmt.Sprintf("%s-%s", ra.Name, ra.Version)
+	}
+
+	// Find addons that need to be created (present in requested, missing from created)
+	addonsToCreate := make([]*v1alpha1.AddOnConfig, 0)
+	for i, requestedName := range requestedVersionedNames {
+		if !slices.Contains(createdVersionedNames, requestedName) {
+			addonsToCreate = append(addonsToCreate, requestedAddOns[i])
+		}
+	}
+
+	// Find addons that need to be deleted (present in created, missing from requested)
+	addonsToDelete := make([]string, 0)
+	for _, createdName := range createdVersionedNames {
+		if !slices.Contains(requestedVersionedNames, createdName) {
+			addonsToDelete = append(addonsToDelete, createdName)
+		}
+	}
+
+	// do deletes first, then creates.
+	if len(addonsToDelete) > 0 {
+		err = handleAddonDelete(ctx, addonC, fc, addonsToDelete)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(addonsToCreate) > 0 {
+		err = handleAddonCreate(ctx, kClient, fc, addonsToCreate)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
IMO it would be cleaner to do the length check at the start of the handleAddonXYZ() functions. That way the function itself handles the empty/nil case and we wouldn't need to worry about checking length before calling them anywhere.